### PR TITLE
If a deleted parent doesn't have a thread anymore, hide it

### PR DESCRIPTION
**Pull request by dangrous**
### Details

Hides the `[Deleted message]` of a parent of a thread if the thread no longer exists. Actually applies in a broader set of circumstances than listed in the linked issue (unless I'm missing something).

It introduces one slight UI weirdness that I'd love your thoughts on, reviewers! Namely, if you delete the last child message of a thread, you are still in that thread, with only the [Deleted message] there, but as soon as you leave that thread disappears. I think that that's fine? At least I'm not sure what else to do that would be better.

Additionally, there is some weirdness with how these tests behave offline, but I checked and it seems to be the same on dev (the thread still is shown in the parent report even if it's deleted) so I'm not worrying about that.

### Fixed Issues

$ https://github.com/Expensify/App/issues/21508
$ https://github.com/Expensify/App/issues/22674

### Tests

Same as QA

- [X] Verify that no errors appear in the JS console

### QA Steps

1. In any chat, add a message
2. Add another message as a thread off of that message
3. Delete the initial message, either from the parent report or the chat
4. Verify it is shown in both places as `[Deleted message]` and that the thread still exists and is accessible
5. Delete the message in the thread
6. Once you move away from the thread, ensure the thread no longer exists in the LHN
7. Ensure that there is no `[Deleted message]` in the parent chat anymore
8. [Additional tests for second issue]
9. Send a new message
10. Open a thread on that message, but don't actually send anything
11. Go back to the main chat
12. Set your connection to Slow 3G
13. Delete that message
14. Ensure you don't see `[Deleted message]` and the chat is fully deleted immediately.

- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
Still having trouble with mWeb Chrome overall, unable to load (on main too)
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/5487802/057fddd5-f878-4176-b76d-3b5e19fb980d

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/5487802/483a4c88-ed03-41c4-adb6-cd2138d0c3af

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/5487802/786ba79b-3753-4402-aa36-338b97e5d4e7

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/5487802/4b05c06e-0f52-4fd1-a224-0265e774a10d

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/5487802/b1128380-c42d-4f0d-8860-92c38cb6c058

</details>

### CODE REVIEW

**Review by mananjadhav — APPROVED** [review]
Change looks fine, testing this out.

**Comment by mananjadhav** [comment]
@dangrous I can see an [issue](https://github.com/Expensify/App/issues/22674) related to `[Deleted Message]` visibility. do we want to tackle it here itself?

Steps:
1. Go to any chat, and send a message
2. Click on `Reply to thread` for the sent message (it should open the thread screen)
3. Do not enter any message and click on the header (`From ...`) to open the original chat
4. Delete the last sent message, and you'll see `[Deleted Message]` momentarily.

**Comment by dangrous** [comment]
Hmmm @mananjadhav yeah that first issue seems like something we can tackle here as well. I think we just need to update the optimistic data to match, let me see what I can do.

As for the second you called out, I'm not seeing that, is there a specific circumstance that made that happen (e.g. slow connection, something like that?). If you inspect those with react dev tools, what are the values of `isDeletedParentAction` and `childVisibleActionCount`?

**Comment by dangrous** [comment]
I added a change that seems to fix that other issue, using hasCommentThread. Seems to work on my end, let me know how it tests!

> Also after refreshing/logging out and logging in I can see these [Deleted Message] messages back.
@mananjadhav I managed to see the behavior you're talking about - but it only happens when I revert the code in this PR... Is it possible you didn't have the branch loaded when you were testing out that other issue?

**Comment by mananjadhav** [comment]
You're right. I think I had switched my branch by then. I'll retest the feature in your branch.

**Comment by dangrous** [comment]
> @dangrous one more scenario, assume before we delete the last message, we copy the thread chat URL, and after deleting the last message we paste and visit the thread URL. At the moment, it loads `[Deleted Message]` chat window. I think this is fine for the expected behavior right? or should it redirect to the parent chat?

Hmmm yeah I think that's probably fine. That way, if there was an old URL somewhere, you'd know what happened (oh, the thread has been deleted) vs. just being redirected without a note. Technically then you could reopen a thread on a deleted message that you otherwise can't see, but I don't see how that would be used, for good or for bad.

**Review by mananjadhav — APPROVED** [review]

**Comment by mananjadhav** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/3069065/071905b9-6447-4f1e-ad79-8bbd8f7a9fc5

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/3069065/c95a3aa2-dd2d-474e-aa12-8242e6befe90

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/3069065/d514367d-05d0-4a1e-b79e-56eb2eb29c20

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/3069065/fb95a766-b3a4-40d1-80db-f96b7bddef98

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/3069065/8d6dc154-a547-43e1-8b29-a06731e79673

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/3069065/6fd3e732-c6ae-4900-b0e6-4bede32c8628

</details>


-----

Thanks for the answers @dangrous. Can you rerun the lint action? It looks failed. @yuwenmemon All yours.

**Comment by dangrous** [comment]
Yeah lint is not happy - seems like it's not just this PR, discussing [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1689189930857949)

**Comment by dangrous** [comment]
woo lint is good to go

**Review by srikarparsi — APPROVED** [review]
lgtm!

**Comment by dangrous** [comment]
So I think Melvin double-assigned an internal reviewer here, I don't think we need you @yuwenmemon but if you'd like to join in, have at it!

**Comment by srikarparsi** [comment]
Cool! I'll just go ahead and merge this since it seems pretty straightforward

File: src/libs/ReportActionsUtils.js [hunk 1]
```diff
@@ -90,7 +90,7 @@ function isReportPreviewAction(reportAction) {
  * @returns {Boolean}
  */
 function hasCommentThread(reportAction) {
-    return lodashGet(reportAction, 'childType', '') === CONST.REPORT.TYPE.CHAT;
+    return lodashGet(reportAction, 'childType', '') === CONST.REPORT.TYPE.CHAT && lodashGet(reportAction, 'childVisibleActionCount', 0) > 0;
```
[the remark this answers is not in the document]
Comment (dangrous) [reply src/libs/ReportActionsUtils.js:93]: Hm, that's interesting. Can you bring this up in Slack? It's probably not ideal behavior but I honestly can't think of what would be better. Like we'd still want to be able to access that content, so maybe it's fine? Do you think we should keep `[Deleted message]` all the way down until there's a real thread? It seems a bit confusing.

**Pull request merged**